File: source-connector/src/main/java/sourceconnector/repository/file/decompressor/DecompressorSelector.java
package sourceconnector.repository.file.decompressor;

import lombok.RequiredArgsConstructor;
import offsetmanager.domain.file.FileKey;

@RequiredArgsConstructor
public enum DecompressorSelector {
  GZIP(new GzipDecompressor()),
  ZIP(new ZipDecompressor()),
  ZSTD(new ZstdDecompressor()),
  NONE(new NoneDecompressor());

  private final Decompressor decompressor;

  public static Decompressor select(FileKey fileKey) {
    for (DecompressorSelector selector : values()) {
      if (selector.decompressor.supports(fileKey)) {
        return selector.decompressor;
      }
    }
    throw new IllegalArgumentException("No supported decompressor found for file: " + fileKey.get());
  }

}

File: source-connector/src/main/kotlin/sourceconnector/repository/file/decompressor/DecompressorSelector.kt
package sourceconnector.repository.file.decompressor

import offsetmanager.domain.file.FileKey

enum class DecompressorSelector(private val decompressor: Decompressor) {
  GZIP(GzipDecompressor()),
  ZIP(ZipDecompressor()),
  ZSTD(ZstdDecompressor()),
  NONE(NoneDecompressor()); // Semicolon is needed to separate enum constants from other members

  companion object {
    /**
     * Selects the first decompressor that supports the given file key.
     * @throws IllegalArgumentException if no supported decompressor is found.
     */
    @JvmStatic
    fun select(fileKey: FileKey): Decompressor {
      return entries
        .firstOrNull { selector -> selector.decompressor.supports(fileKey) }
        ?.decompressor
        ?: throw IllegalArgumentException("No supported decompressor found for file: ${fileKey.get()}")
    }
  }
}
